test/e2e/apps/job.go

```diff
@@ -344,6 +344,48 @@ var _ = SIGDescribe("Job", func() {
 		}
 	})
 
+	ginkgo.It("should recreate pods only after they have failed if pod replacement policy is set to Failed", func(ctx context.Context) {
+		ginkgo.By("Creating a job")
+		job := e2ejob.NewTestJob("notTerminate", "pod-recreate-failed", v1.RestartPolicyNever, 1, 1, nil, backoffLimit)
+		job.Spec.PodReplacementPolicy = ptr.To(batchv1.Failed)
+		job.Spec.Template.Spec.Containers[0].Image = "gcr.io/k8s-staging-perf-tests/sleep"
+		job.Spec.Template.Spec.Containers[0].Args = []string{"1m", "-termination-grace-period", "1m"}
+		job, err := e2ejob.CreateJob(ctx, f.ClientSet, f.Namespace.Name, job)
+		framework.ExpectNoError(err, "failed to create job in namespace: %s", f.Namespace.Name)
+
+		err = e2ejob.WaitForJobPodsRunning(ctx, f.ClientSet, f.Namespace.Name, job.Name, 1)
+		framework.ExpectNoError(err, "failed to wait for job pod to become running in namespace: %s", job.Name, f.Namespace.Name)
+
+		ginkgo.By("Deleting job pod")
+		pods, err := e2ejob.GetJobPods(ctx, f.ClientSet, f.Namespace.Name, job.Name)
+		framework.ExpectNoError(err, "failed to get pod list for job %s in namespace: %s", job.Name, f.Namespace.Name)
+
+		framework.ExpectNoError(e2epod.DeletePodsWithGracePeriod(ctx, f.ClientSet, pods.Items, 60), "failed to delete pods in namespace: %s", f.Namespace.Name)
+
+		ginkgo.By("Ensuring pod does not get recreated while it is in terminating state")
+		job, err = e2ejob.GetJob(ctx, f.ClientSet, f.Namespace.Name, job.Name)
+		framework.ExpectNoError(err, "failed to get job in namespace: %s", f.Namespace.Name)
+		gomega.Expect(job.Status.Active).To(gomega.Equal(int32(0)), "expected job to have 0 active pods")
+		gomega.Expect(job.Status.Failed).To(gomega.Equal(int32(0)), "expected job to have 0 failed pods")
+		gomega.Expect(job.Status.Terminating).To(gomega.Equal(ptr.To[int32](1)), "expected job to have 1 terminating pod")
```

Comment:
You should check all of this conditions with a single `WaitForJobState`, to give time to the job controller to update status after you issue the Pod deletion.